File: Exam/doublell.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* prev;
    struct Node* next;
};

struct Node* head = NULL;

// Display elements
void display() {
    struct Node* temp = head;
    if (temp == NULL) {
        printf("List is empty.\n");
        return;
    }
    printf("List elements: ");
    while (temp != NULL) {
        printf("%d ", temp->data);
        temp = temp->next;
    }
    printf("\n");
}

// Count nodes
int countNodes() {
    int count = 0;
    struct Node* temp = head;
    while (temp != NULL) {
        count++;
        temp = temp->next;
    }
    return count;
}

// Add at beginning
void addAtBeginning(int data) {
    struct Node* new_node = (struct Node*)malloc(sizeof(struct Node));
    new_node->data = data;
    new_node->prev = NULL;
    new_node->next = head;

    if (head != NULL) {
        head->prev = new_node;
    }
    head = new_node;
    printf("Added %d at the beginning.\n", data);
}

// Add at end
void addAtEnd(int data) {
    struct Node* new_node = (struct Node*) malloc(sizeof(struct Node));
    new_node->data = data;
    new_node->next = NULL;

    if (head == NULL) {
        new_node->prev = NULL;
        head = new_node;
        printf("Added %d as the first element.\n", data);
        return;
    }

    struct Node* temp = head;
    while (temp->next != NULL) {
        temp = temp->next;
    }
    temp->next = new_node;
    new_node->prev = temp;
    printf("Added %d at the end.\n", data);
}

// Add at position
void addAtPosition(int data, int pos) {
    int length = countNodes();
    if (pos < 1 || pos > length + 1) {
        printf("Invalid position.\n");
        return;
    }
    if (pos == 1) {
        addAtBeginning(data);
        return;
    }
    if (pos == length + 1) {
        addAtEnd(data);
        return;
    }

    struct Node* new_node = (struct Node*)malloc(sizeof(struct Node));
    new_node->data = data;

    struct Node* temp = head;
    int i;
    for (i = 1; i < pos - 1; i++) {
        temp = temp->next;
    }

    new_node->next = temp->next;
    new_node->prev = temp;
    if (temp->next != NULL)
        temp->next->prev = new_node;
    temp->next = new_node;

    printf("Added %d at position %d.\n", data, pos);
}

// Search element
void search(int key) {
    struct Node* temp = head;
    int pos = 1;
    while (temp != NULL) {
        if (temp->data == key) {
            printf("Element %d found at position %d.\n", key, pos);
            return;
        }
        temp = temp->next;
        pos++;
    }
    printf("Element %d not found in the list.\n", key);
}

// Delete from beginning
void deleteFromBeginning() {
    if (head == NULL) {
        printf("List is empty.\n");
        return;
    }
    struct Node* temp = head;
    head = head->next;
    if (head != NULL) {
        head->prev = NULL;
    }
    printf("Deleted element %d from beginning.\n", temp->data);
    free(temp);
}

// Delete from end
void deleteFromEnd() {
    if (head == NULL) {
        printf("List is empty.\n");
        return;
    }
    struct Node* temp = head;
    if (temp->next == NULL) { // only one node
        head = NULL;
        printf("Deleted element %d from end.\n", temp->data);
        free(temp);
        return;
    }
    while (temp->next != NULL) {
        temp = temp->next;
    }
    temp->prev->next = NULL;
    printf("Deleted element %d from end.\n", temp->data);
    free(temp);
}

// Delete from position
void deleteFromPosition(int pos) {
    int length = countNodes();
    if (pos < 1 || pos > length) {
        printf("Invalid position.\n");
        return;
    }
    if (pos == 1) {
        deleteFromBeginning();
        return;
    }
    if (pos == length) {
        deleteFromEnd();
        return;
    }

    struct Node* temp = head;
    int i;
    for (i = 1; i < pos; i++) {
        temp = temp->next;
    }
    temp->prev->next = temp->next;
    if (temp->next != NULL)
        temp->next->prev = temp->prev;

    printf("Deleted element %d from position %d.\n", temp->data, pos);
    free(temp);
}

int main() {
    int choice, data, pos;
    while (1) {
        printf("\n--- Doubly Linked List Menu ---\n");
        printf("1. Display\n");
        printf("2. Add at Beginning\n");
        printf("3. Add at End\n");
        printf("4. Add at Position\n");
        printf("5. Search Element\n");
        printf("6. Count Nodes\n");
        printf("7. Delete from Beginning\n");
        printf("8. Delete from End\n");
        printf("9. Delete from Position\n");
        printf("10. Exit\n");
        printf("Enter your choice: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                display();
                break;
            case 2:
                printf("Enter data to add at beginning: ");
                scanf("%d", &data);
                addAtBeginning(data);
                break;
            case 3:
                printf("Enter data to add at end: ");
                scanf("%d", &data);
                addAtEnd(data);
                break;
            case 4:
                printf("Enter data to add: ");
                scanf("%d", &data);
                printf("Enter position: ");
                scanf("%d", &pos);
                addAtPosition(data, pos);
                break;
            case 5:
                printf("Enter element to search: ");
                scanf("%d", &data);
                search(data);
                break;
            case 6:
                printf("Number of nodes: %d\n", countNodes());
                break;
            case 7:
                deleteFromBeginning();
                break;
            case 8:
                deleteFromEnd();
                break;
            case 9:
                printf("Enter position to delete: ");
                scanf("%d", &pos);
                deleteFromPosition(pos);
                break;
            case 10:
                printf("Exiting program.\n");
                exit(0);
            default:
                printf("Invalid choice. Try again.\n");
        }
    }
    return 0;
}

File: Exam/doublell.py
class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None

    # Display the list
    def display(self):
        if not self.head:
            print("List is empty.")
            return
        temp = self.head
        print("List elements:", end=" ")
        while temp:
            print(temp.data, end=" ")
            temp = temp.next
        print()

    # Count nodes
    def count_nodes(self):
        count = 0
        temp = self.head
        while temp:
            count += 1
            temp = temp.next
        print("Number of nodes:", count)
        return count

    # Add at beginning
    def add_at_beginning(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
        else:
            new_node.next = self.head
            self.head.prev = new_node
            self.head = new_node
        print(f"Added {data} at the beginning.")

    # Add at end
    def add_at_end(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
        else:
            temp = self.head
            while temp.next:
                temp = temp.next
            temp.next = new_node
            new_node.prev = temp
        print(f"Added {data} at the end.")

    # Add at position
    def add_at_position(self, data, pos):
        if pos <= 0:
            print("Position should be >= 1")
            return
        new_node = Node(data)
        if pos == 1:
            self.add_at_beginning(data)
            return
        temp = self.head
        for i in range(pos - 2):
            if temp is None:
                print("Position out of bounds")
                return
            temp = temp.next
        if temp is None:
            print("Position out of bounds")
            return
        new_node.next = temp.next
        if temp.next:
            temp.next.prev = new_node
        temp.next = new_node
        new_node.prev = temp
        print(f"Added {data} at position {pos}.")

    # Search element
    def search(self, key):
        temp = self.head
        pos = 1
        while temp:
            if temp.data == key:
                print(f"Element {key} found at position {pos}.")
                return
            temp = temp.next
            pos += 1
        print(f"Element {key} not found.")

    # Delete from beginning
    def delete_from_beginning(self):
        if not self.head:
            print("List is empty.")
            return
        removed = self.head.data
        self.head = self.head.next
        if self.head:
            self.head.prev = None
        print(f"Deleted element {removed} from beginning.")

    # Delete from end
    def delete_from_end(self):
        if not self.head:
            print("List is empty.")
            return
        temp = self.head
        if temp.next is None:
            removed = temp.data
            self.head = None
            print(f"Deleted element {removed} from end.")
            return
        while temp.next:
            temp = temp.next
        removed = temp.data
        temp.prev.next = None
        print(f"Deleted element {removed} from end.")

    # Delete from position
    def delete_from_position(self, pos):
        if not self.head:
            print("List is empty.")
            return
        if pos <= 0:
            print("Position should be >= 1")
            return
        if pos == 1:
            self.delete_from_beginning()
            return
        temp = self.head
        for i in range(pos - 1):
            if temp is None:
                print("Position out of bounds")
                return
            temp = temp.next
        if temp is None:
            print("Position out of bounds")
            return
        removed = temp.data
        if temp.next:
            temp.next.prev = temp.prev
        if temp.prev:
            temp.prev.next = temp.next
        print(f"Deleted element {removed} from position {pos}.")

def main():
    dll = DoublyLinkedList()
    while True:
        print("\nMenu:")
        print("1. Display")
        print("2. Add at Beginning")
        print("3. Add at End")
        print("4. Add at Position")
        print("5. Search Element")
        print("6. Count Nodes")
        print("7. Delete from Beginning")
        print("8. Delete from End")
        print("9. Delete from Position")
        print("10. Exit")
        choice = int(input("Enter your choice: "))

        if choice == 1:
            dll.display()
        elif choice == 2:
            data = int(input("Enter data: "))
            dll.add_at_beginning(data)
        elif choice == 3:
            data = int(input("Enter data: "))
            dll.add_at_end(data)
        elif choice == 4:
            data = int(input("Enter data: "))
            pos = int(input("Enter position: "))
            dll.add_at_position(data, pos)
        elif choice == 5:
            key = int(input("Enter element to search: "))
            dll.search(key)
        elif choice == 6:
            dll.count_nodes()
        elif choice == 7:
            dll.delete_from_beginning()
        elif choice == 8:
            dll.delete_from_end()
        elif choice == 9:
            pos = int(input("Enter position to delete: "))
            dll.delete_from_position(pos)
        elif choice == 10:
            print("Exiting program.")
            break
        else:
            print("Invalid choice. Please try again.")

if __name__ == "__main__":
    main()
